lib/model/message_list.dart

```diff
@@ -341,6 +341,38 @@ class MessageListView with ChangeNotifier, _MessageSequence {
     notifyListeners();
   }
 
+  void maybeUpdateMessageFlags(UpdateMessageFlagsEvent event) {
+    final List<Message> messagesToProcess;
+    if (event.all) {
+      messagesToProcess = messages;
+    } else {
+      messagesToProcess = [];
+      for (final messageId in event.messages) {
+        // TODO(perf) could be more efficient?
+        final index = _findMessageWithId(messageId);
+        if (index != -1) {
+          messagesToProcess.add(messages[index]);
```

Comment:
The algorithmic comment I'd make on a different axis, though, is that I'd rather not build up this list. Instead, this loop can go ahead and do the add/remove of flags itself. Can have a boolean local variable that says if it's add or remove (from looking once at the type of the event), and condition on that inside the loop.